In [1]:
import pandas as pd
import numpy as np

<hr style="border-top: 10px groove darkorange; margin-top: 1px; margin-bottom: 1px"></hr>

# Tidy Data

___

## Attendance Data

Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [5]:
att_df = pd.read_csv('untidy-data/attendance.csv', index_col=0)
att_df

2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
Sally          P          T          T          H          P          A   
Jane           A          P          T          T          T          T   
Billy          A          T          A          A          H          T   
John           P          T          H          P          P          T   

      2018-01-07 2018-01-08  
Sally          T          T  
Jane           A          T  
Billy          P          T  
John           P          P

In [11]:
att_df = att_df.reset_index().rename(columns={'index': 'name'})
att_df

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [13]:
# melt the data; one observation is one name and one day.

att_df = att_df.melt(id_vars='name')
att_df.sample(5)

name    variable value
14  Billy  2018-01-04     A
30  Billy  2018-01-08     T
10  Billy  2018-01-03     A
17   Jane  2018-01-05     T
24  Sally  2018-01-07     T

In [15]:
# replace the categorical values in value with the numeric values

att_df['value'] = att_df.value.map({'P': 1, 'H': .5, 'T': .9, 'A': 0})
att_df.head()

name    variable  value
0  Sally  2018-01-01    1.0
1   Jane  2018-01-01    0.0
2  Billy  2018-01-01    0.0
3   John  2018-01-01    1.0
4  Sally  2018-01-02    0.9

In [16]:
# calculate an overall average for each person.

att_df.groupby('name').value.mean()

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: value, dtype: float64

___

In [51]:
att_df = pd.read_csv('untidy-data/attendance.csv')
att_df = att_df.rename(columns={'Unnamed: 0': 'student'})
att_df

student 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0   Sally          P          T          T          H          P          A   
1    Jane          A          P          T          T          T          T   
2   Billy          A          T          A          A          H          T   
3    John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [52]:
att_df = att_df.melt(id_vars='student')
att_df.head()

student    variable value
0   Sally  2018-01-01     P
1    Jane  2018-01-01     A
2   Billy  2018-01-01     A
3    John  2018-01-01     P
4   Sally  2018-01-02     T

In [53]:
att_df['grade'] = att_df.value.map({'P': 1, 'H': .5, 'T': .9, 'A': 0})
att_df.head()

student    variable value  grade
0   Sally  2018-01-01     P    1.0
1    Jane  2018-01-01     A    0.0
2   Billy  2018-01-01     A    0.0
3    John  2018-01-01     P    1.0
4   Sally  2018-01-02     T    0.9

___

## Coffee Levels

### Read the coffee_levels.csv file.

In [31]:
coffee_df = pd.read_csv('untidy-data/coffee_levels.csv', index_col=0)
coffee_df.sample(5)

coffee_carafe  coffee_amount
hour                             
13               y       0.997464
16               y       0.544676
14               y       0.058361
14               z       0.864464
11               x       0.335533

### Transform the data so that each carafe is in it's own column.

In [54]:
coffee_pivot = coffee_df.reset_index().pivot_table(values='coffee_amount', index='hour', columns='coffee_carafe')
coffee_pivot.sample(5)

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
15             0.215043  0.144644  0.436364
10             0.843279  0.023163  0.144928
13             0.310711  0.997464  0.398520
16             0.183891  0.544676  0.280621

### Is this the best shape for the data?

- This is a decent way to view the data, but it was tidier to start because you can use a `.groupby()` function, and there is one measure per column.

In [34]:
# I like this view, as well.

coffee_long = coffee_df.reset_index().pivot(index='coffee_carafe', columns='hour')
coffee_long

coffee_amount                                                    \
hour                     8         9         10        11        12        13   
coffee_carafe                                                                   
x                  0.816164  0.451018  0.843279  0.335533  0.898291  0.310711   
y                  0.189297  0.521502  0.023163  0.235529  0.017009  0.997464   
z                  0.999264  0.915990  0.144928  0.311495  0.771947  0.398520   

                                                       
hour                 14        15        16        17  
coffee_carafe                                          
x              0.507288  0.215043  0.183891  0.391560  
y              0.058361  0.144644  0.544676  0.594126  
z              0.864464  0.436364  0.280621  0.436677

___

## Cake Recipes

### Read the cake_recipes.csv data. 

- This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [61]:
cake = pd.read_csv('untidy-data/cake_recipes.csv')
cake

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

In [62]:
rec_pos = cake['recipe:position'].str.split(':', expand=True).rename(columns={0: 'recipe', 1: 'position'})
rec_pos

recipe position
0      a   bottom
1      a      top
2      b   bottom
3      b      top
4      c   bottom
5      c      top
6      d   bottom
7      d      top

In [65]:
cake = pd.concat([cake, rec_pos], axis=1).drop(columns='recipe:position')
cake

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom
5  71.306308  82.795477  92.098049  53.960273      c      top
6  52.799753  58.670419  51.747686  56.183110      d   bottom
7  96.873178  76.101363  59.571620  50.971626      d      top

In [74]:
cake_melt = cake.melt(id_vars=['recipe', 'position'], var_name='temp', value_name='score')
cake_melt.head()

recipe position temp      score
0      a   bottom  225  61.738655
1      a      top  225  51.709751
2      b   bottom  225  57.095320
3      b      top  225  82.455004
4      c   bottom  225  96.470207

### Which recipe, on average, is the best? 

- recipe b

In [76]:
cake_melt.groupby('recipe').score.mean().sort_values(ascending=False)

recipe
b    76.736074
c    75.874748
a    63.922201
d    62.864844
Name: score, dtype: float64

### Which oven temperature, on average, produces the best results? 

- 275

In [78]:
cake_melt.groupby('temp').score.mean().sort_values(ascending=False)

temp
275    74.886754
225    71.306022
300    66.627655
250    66.577437
Name: score, dtype: float64

### Which combination of recipe, rack position, and temperature gives the best result? 

- recipe b, bottom rack, 300 degrees

In [82]:
cake_melt.groupby(['recipe', 'position', 'temp']).score.mean().sort_values(ascending=False).head()

recipe  position  temp
b       bottom    300     99.248541
a       bottom    300     98.786784
b       top       275     98.594881
d       top       225     96.873178
c       bottom    225     96.470207
Name: score, dtype: float64